lib/http2Request.js

```diff
@@ -0,0 +1,45 @@
+/*!
+ * express
+ * Copyright(c) 2009-2013 TJ Holowaychuk
+ * Copyright(c) 2013 Roman Shtylman
+ * Copyright(c) 2014-2015 Douglas Christopher Wilson
+ * MIT Licensed
+ */
+
+'use strict';
+
+/**
+ * Module dependencies.
+ * @private
+ */
+var http2 = require('http2');
+var http2Constants = http2.constants;
+var decorator = require('./requestDecorator');
+var defineGetter = require('./utils').defineGetter;
+var http2Req = decorator(Object.create(http2.Http2ServerRequest.prototype));
+var parse = require('parseurl');
+/**
+ * Override default http2 path to return only pathname
+ * as it returns full url
+ * Gets the full URL from the headers
+ * Creates a mock object for parse to cache on and read from
+ * as url in http2 refers to this.path which will results in an error
+ */
+defineGetter(http2Req, 'path', function path() {
+  var path;
+  var headers = this.headers;
+  if (headers === undefined) {
+    return '';
+  }
+  path = headers[http2Constants.HTTP2_HEADER_PATH];
+  this.parsedObject = this.parsedObject || {url: path};

```

Comment:
The req.path should reflect any changes made to req.url over thr course of the request (for example middleware cut thr matched part). This is accidentally caching the first access value.

Reply:
It does not, if the url is set, it will rewrite the path, which in turn will rewrite the url on mock object and cause parse url to rerun.

Reply:
You are correct, I see the object alteration below, sorry! It would be awesome if it could just use the same cache that the routing used, but if this is the only way, then we'll just have to live with the perf hit req.path will get.

Reply:
The only performance hit (to the best of my understanding is the creation of the new object - otherwise it uses the same caching mechanism and everything works exactly the same way, instead of operation on the req object though it operates on a new (only created once) object, and the cache is set on it.

Reply:
That's not the case here. The router itself in Express is prepopulating the cache, but this req.path is using a separate cache, requiring a double parse. Since the movement through middleware will keep altering req.url, the internal prepopulation the router is doing for that is not getting picked up here. For example, the router will populate the cache and then in a middleware you call req.path it has to parse again instead of reusing the cache. Then the request moves on, the router alters req.url and saves the new parse in the cache. The call to req.path again requires yet another reparse since it uses a separate cache, etc. etc.

Reply:
I found a better solution, in retrospect my solution was bad :)

I changed it to basically work the same, except that I rewrite the getter for url to give the raw path (avoiding the endless loop on get path / url until it explodes).

Basically what happens in http2:
get path -> overridden by the decorator to return parse object pathname
set path -> sets the full path in the headers
get url -> brings the full path from headers (instead of this.path as in the unedited Http2ServerRequest)
set url -> sets the path